### ISSUE TITLE
你好，能帮忙指点下吗? 改成3d 后 pos_mat_small 维度不是Scale x  Scale x Scale x 3的维度？ h_offset这需要改吗？

### ISSUE THREAD

**Issue by NewboySoCool** [XuecaiHu/Meta-SR-Pytorch#80]
`def input_matrix_wpn_new(inH, inW, inD, scale, add_scale=True): #10, 10, 10, 2.0
    '''
    inH, inW: the size of the feature maps
    scale: is the upsampling times
    '''
    #计算上采样后的输出大小
    outH, outW, outD = int(scale * inH), int(scale * inW), int(scale * inD)
    #### mask records which pixel is invalid, 1 valid or o invalid
    #### h_offset and w_offset caculate the offset to generate the input matrix
    scale_int = int(math.ceil(scale))

    h_offset = torch.ones(inH, scale_int, 1) #10x2x1  10x2x2x1
    mask_h = torch.zeros(inH, scale_int, 1)

    w_offset = torch.ones(1, inW, scale_int)
    mask_w = torch.zeros(1, inW, scale_int)

    d_offset = torch.ones(scale_int, 1, inD)
    mask_d = torch.zeros(scale_int, 1, inD)

    ####projection  coordinate  and caculate the offset, 如果改成3d这里加一个 d_project_coord
    #h
    h_project_coord = torch.arange(0, outH, 1).mul(1.0 / scale)  #[0, 0.5, 1, 3/2, 4/2, 5/2, .... 9/2]
    int_h_project_coord = torch.floor(h_project_coord)           #再向下取整 [0, 0, 1.0, 1.0, 2.0, 2.0, 3.0, 3.0, 4.0, 4.0] #浮点数

    offset_h_coord = h_project_coord - int_h_project_coord       #[0, 0.5,  0, 0.5,  0, 0.5,  0, 0.5,   0, 0.5]
    int_h_project_coord = int_h_project_coord.int()              #[0, 0, 1, 1, 2, 2 , 3, 3, 4, 4] int类型

    #w 同理
    w_project_coord = torch.arange(0, outW, 1).mul(1.0 / scale)
    int_w_project_coord = torch.floor(w_project_coord)

    offset_w_coord = w_project_coord - int_w_project_coord
    int_w_project_coord = int_w_project_coord.int()

    #D
    d_project_coord = torch.arange(0, outD, 1).mul(1.0 / scale)
    int_d_project_coord = torch.floor(d_project_coord)

    offset_d_coord = d_project_coord - int_d_project_coord
    int_d_project_coord = int_d_project_coord.int()

    ####flag for   number for current coordinate LR image, 改成3d这里应该也得加
    flag = 0
    number = 0
    #int_h_project_coord =  [0, 0, 1, 1, 2, 2 , 3, 3, 4, 4] int类型
    for i in range(outH):
        if int_h_project_coord[i] == number:
            h_offset[int_h_project_coord[i], flag, 0] = offset_h_coord[i]  # h_offset = 10x2x1
            mask_h[int_h_project_coord[i], flag, 0] = 1  # mask_h = 10x2x1
            flag += 1
        else:
            h_offset[int_h_project_coord[i], 0, 0] = offset_h_coord[i]
            mask_h[int_h_project_coord[i], 0, 0] = 1
            number += 1
            flag = 1

    # W
    flag = 0
    number = 0
    for i in range(outW):
        if int_w_project_coord[i] == number:
            w_offset[0, int_w_project_coord[i], flag] = offset_w_coord[i]
            mask_w[0, int_w_project_coord[i], flag] = 1
            flag += 1
        else:
            w_offset[0, int_w_project_coord[i], 0] = offset_w_coord[i]
            mask_w[0, int_w_project_coord[i], 0] = 1
            number += 1
            flag = 1
    #D
    flag = 0
    number = 0
    for i in range(outD):
        if int_d_project_coord[i] == number:
            d_offset[flag, 0, int_d_project_coord[i]] = offset_d_coord[i]
            mask_d[flag, 0, int_d_project_coord[i]] = 1
        else:
            d_offset[0, 0, int_d_project_coord[i]] = offset_d_coord[i]
            mask_d[0, 0, int_d_project_coord[i]] = 1
            number += 1
            flag = 1

    ## the size is scale_int* inH* (scal_int*inW),   h_offset 10x2x1  w_offset 1x10x2  d_offset 2x1x10
    h_offset_coord = torch.cat([h_offset] * (scale_int * inW * scale_int * inD), 2).view(-1, scale_int * inW, scale_int * inD, 1)
    print('h_offset_coord', h_offset_coord.shape)
    w_offset_coord = torch.cat([w_offset] * (scale_int * inH * scale_int * inD), 0).view(scale_int * inH, -1, scale_int * inD, 1)
    print('w_offset_coord', w_offset_coord.shape)
    d_offset_coord = torch.cat([d_offset] * (scale_int * inH * scale_int * inW), 1).view(scale_int * inH, scale_int * inW, -1, 1)
    print('w_offset_coord', w_offset_coord.shape) #20x20x20x1

    ####
    mask_h = torch.cat([mask_h] * (scale_int * inW * scale_int * inD), 2).view(-1, scale_int * inW, scale_int * inD, 1)
    print('mask_h', mask_h.shape) #20x20x20x1
    mask_w = torch.cat([mask_w] * (scale_int * inH * scale_int * inD), 0).view(scale_int * inH, -1, scale_int * inD, 1)
    print('mask_w', mask_w.shape) #20x20x20x1
    mask_d = torch.cat([mask_d] * (scale_int * inH * scale_int * inW), 1).view(scale_int * inH, scale_int * inW, -1, 1)
    print('mask_w', mask_w.shape) #20x20x20x1

    #前面的操作把维度=1放在最后，最前面是超分后的结果大小。
    pos_mat = torch.cat((h_offset_coord, w_offset_coord, d_offset_coord), 3)
    print('pos_mat', pos_mat.shape) #应该得到20x20x20x3

    mask_mat = torch.sum(torch.cat((mask_h, mask_w, mask_d), 3), 3).view(scale_int * inH, scale_int * inW, scale_int * inD)
    print('mask_mat', mask_mat.shape)  #20x20x20

    mask_mat = mask_mat.eq(3) #与2进行比较，元素>2返回true，生成掩码

    i = 1      #3d pos_mat 20x20x20x1, 用k 表示d
    h, w, d, _ = pos_mat.size()  #20x20x20x3
    while (pos_mat[i][0][0][0] >= 1e-6 and i < h):
        i = i + 1

    j = 1
    # pdb.set_trace()
    h, w, d, _ = pos_mat.size()
    while (pos_mat[0][j][0][1] >= 1e-6 and j < w):
        j = j + 1

    k = 1
    # pdb.set_trace()
    h, w, d, _ = pos_mat.size()
    while (pos_mat[0][0][k][2] >= 1e-6 and j < d):
        k = k + 1
    print("k", k)

    pos_mat_small = pos_mat[0:i, 0:j, 0:k, :]  #
    print('pos_mat_small', pos_mat_small.shape) #是 2x2x2x3才对

    pos_mat_small = pos_mat_small.contiguous().view(1, -1, 3) #  1x8x3 这是 1x(scale**3)x3
    print('pos_mat_small', pos_mat_small.shape)
    if add_scale:
        scale_mat = torch.zeros(1, 1)
        scale_mat[0, 0] = 1.0 / scale
        scale_mat = torch.cat([scale_mat] * (pos_mat_small.size(1)), 0)
        print('scale_mat', scale_mat.shape) #8x1才对  (scale**3) x 1才对
        pos_mat_small = torch.cat((scale_mat.view(1, -1, 1), pos_mat_small), 2) #1x4x1 跟 1x4x2 合并 = 1x4x3

    return pos_mat_small, mask_mat

`

**Comment by XuecaiHu** [comment]
`
import torch  
import math 
import pdb
def  OneDimLocationProject(inSize,scale,dim,dim_num=3): 
    '''
    inSize: inH or inW or inD
    scale: super resolution up times 2x or 3x or 3.5 and others
    '''    
    outSize = int(scale*inSize)   
    scale_int = int(math.ceil(scale))
    
    scale_index = torch.arange(0,outSize, 1).float().mul(1.0/scale)
    int_scale_index = torch.floor(scale_index)
    int_scale_index = int_scale_index.int()
    #offset = scale_index - int_scale_index
    mask_mat = torch.zeros(inSize,  scale_int)
    offset_mat = torch.zeros(inSize,  scale_int)
    ####flag for   number for current coordinate LR image
    flag = 0
    number = 0
    for i in range(outSize):
        if int_scale_index[i] == number:
            offset_mat[int_scale_index[i], flag] = scale_index[i]- int_scale_index[i]
            mask_mat[int_scale_index[i], flag] = 1
            flag += 1
        else:
            offset_mat[int_scale_index[i], 0] = scale_index[i]- int_scale_index[i]
            mask_mat[int_scale_index[i], 0] = 1
            number += 1
            flag = 1 
    
    size_list=[1 for i in range(dim_num+1)]
    size_list[dim]= inSize*scale_int
    torch_size = torch.Size(size_list)
    offset_mat = offset_mat.view(torch_size)
    mask_mat = mask_mat.view(torch_size)    
         
    return offset_mat, mask_mat

def input_matrix_wpn(inH, inW,inD, scale, add_scale=True):
    '''
    inH, inW: the size of the feature maps
    scale: is the upsampling times
    output1: offset_mat with size (outH_int,outW_int,outD_int,3)
    output2: mask_mat with size (outH_int,outW_int,outD_int,1)
    '''
    outH, outW,outD = int(scale*inH), int(scale*inW),int(scale*inD)
    scale_int = int(math.ceil(scale))
    outH_int, outW_int, outD_int = scale_int*inH,scale_int*inW,scale_int*inD
    #### mask records which pixel is invalid, 1 valid or o invalid
    #### h_offset and w_offset caculate the offset to generate the input matrix
    h_offset,h_mask = OneDimLocationProject(inH,scale,0)
    w_offset,w_mask = OneDimLocationProject(inW,scale,1)
    d_offset,d_mask = OneDimLocationProject(inD,scale,2)
       
    pdb.set_trace()
    ###repeat all tensor with size outH_int, outW_int, outD_int
    h_offset_re,h_mask_re = h_offset.repeat(1,outW_int,outD_int,1),h_mask.repeat(1,outW_int,outD_int,1)
    w_offset_re,w_mask_re = w_offset.repeat(outH_int,1,outD_int,1),w_mask.repeat(outH_int,1,outD_int,1)
    d_offset_re,d_mask_re = d_offset.repeat(outH_int,outW_int,1,1),d_mask.repeat(outH_int,outW_int,1,1)
    
    
    pos_mat = torch.cat([h_offset_re,w_offset_re,d_offset_re],3)
    mask_mat = h_mask_re + w_mask_re + d_mask_re
    scale_int = int(math.ceil(scale))
    h_offset = torch.ones(inH, scale_int,scale_int, 1)
    mask_h = torch.zeros(inH,  scale_int,scale_int,1)
    w_offset = torch.ones(1, inW, scale_int,scale_int)
    mask_w = torch.zeros(1, inW, scale_int,scale_int)
    mask_mat = mask_mat.eq(3)
    
    if add_scale:
        scale_mat = torch.zeros(1,1)
        scale_mat[0,0] = 1.0/scale
        scale_mat = scale_mat.repeat(outH_int,outW_int,outD_int,1)
        pos_mat = torch.cat([pos_mat,scale_mat],3)

    return pos_mat, mask_mat.squeeze()
`

**Comment by XuecaiHu** [comment]
`
def  OneDimLocationProject(inSize,scale,dim,dim_num=3): 
    '''
    inSize: inH or inW or inD
    scale: super resolution up times 2x or 3x or 3.5 and others
    '''    
    outSize = int(scale*inSize)   
    scale_int = int(math.ceil(scale))
    
    scale_index = torch.arange(0,outSize, 1).float().mul(1.0/scale)
    int_scale_index = torch.floor(scale_index)
    int_scale_index = int_scale_index.int()
    #offset = scale_index - int_scale_index
    mask_mat = torch.zeros(inSize,  scale_int)
    offset_mat = torch.zeros(inSize,  scale_int)
    ####flag for   number for current coordinate LR image
    flag = 0
    number = 0
    for i in range(outSize):
        if int_scale_index[i] == number:
            offset_mat[int_scale_index[i], flag] = scale_index[i]- int_scale_index[i]
            mask_mat[int_scale_index[i], flag] = 1
            flag += 1
        else:
            offset_mat[int_scale_index[i], 0] = scale_index[i]- int_scale_index[i]
            mask_mat[int_scale_index[i], 0] = 1
            number += 1
            flag = 1 
    
    size_list=[1 for i in range(dim_num+1)]
    size_list[dim]= inSize*scale_int
    torch_size = torch.Size(size_list)
    offset_mat = offset_mat.view(torch_size)
    mask_mat = mask_mat.view(torch_size)    
         
    return offset_mat, mask_mat
`

**Comment by XuecaiHu** [comment]
[meta-sr_3d.txt](https://github.com/XuecaiHu/Meta-SR-Pytorch/files/11744824/meta-sr_3d.txt)

**Comment by NewboySoCool** [comment]
> [meta-sr_3d.txt](https://github.com/XuecaiHu/Meta-SR-Pytorch/files/11744824/meta-sr_3d.txt)
非常感谢，我这就试下！感谢

**Comment by NewboySoCool** [comment]
您好，后面那部分2d卷积改成3d卷积底层的更改，能帮助看下嘛？实在没改明白，一到这我就晕了。 unfold函数只针对 NxCxHxW的数据以及在往下该怎么改呢？请求帮助，感谢！

‘ 

class MetaUpSampler(nn.Module):
def __init__(self, n_feats, n_colors, kernel_size):
super(MetaUpSampler, self).__init__()
self.P2W = Pos2Weight(inC=n_feats, outC=n_colors)
self.outC = n_colors
self.inC = n_feats
self.kernel_size = kernel_size

def forward(self, lr_features, sr_scale):
        N, C, inH, inW, inD = lr_features.shape
        device = lr_features.device  #数据所在GPU
        # ## predict local weights
        #先计算矩阵
pos_mat, mask = input_matrix_wpn(inH, inW, inD, sr_scale, False)
print('input_matrix_wpn_new', pos_mat.shape, mask.shape)
        mask = mask.to(device)       #20x20x20
pos_mat = pos_mat.to(device)  # (scale**3) x3  20x20x20x3

local_weight = self.P2W(pos_mat.view(-1, 3)) # pos_mat.view(-1, 3) 8000x3
print('local_weight', local_weight.shape) #8000x5184

up_x = self.repeat_x(lr_features, sr_scale) # lr 1x64x10x10x10   (scale**3) x 64x10x10x10
scale_int =  math.ceil(sr_scale)
print('up_x', up_x.shape)   #8x64x10x10x10
        # N*r*r*r x inC x inH x inW x inD
up_x = up_x.view(N*(scale_int**3), C, inH, inD*inW)

        cols = nn.functional.unfold(up_x, self.kernel_size, padding=1)
print('cols', cols.shape) #N*(r**3) x inC*(r**3) x (inH*inW*inD)

scale_int = math.ceil(sr_scale)
        local_weight = self.repeat_weight(local_weight, scale_int, inH, inW, inD)

        cols = cols.contiguous().view(
            cols.size(0) // (scale_int ** 3), scale_int ** 3, cols.size(1), cols.size(2), 1
).permute(0, 1, 4, 2, 5, 3).contiguous()

        local_weight = local_weight.contiguous().view(
            inH, scale_int, inW, scale_int, inD, scale_int, -1, self.outC
        ).permute(1, 3, 5, 0, 2, 4, 6, 7).contiguous()
        local_weight = local_weight.contiguous().view(
            scale_int ** 3, inH * inW * inD, -1, self.outC
        )

        out = torch.matmul(cols, local_weight).permute(0, 1, 5, 2, 6, 3, 4)
        out = out.contiguous().view(
            N, scale_int, scale_int, scale_int, self.outC, inH, inW, inD
        ).permute(0, 4, 5, 6, 1, 2, 3, 7)
        out = out.contiguous().view(
            N, self.outC, scale_int * inH, scale_int * inW, scale_int * inD
        )

# ## mask
out = torch.masked_select(out, mask)
        out = out.contiguous().view(
            N, self.outC, int(sr_scale * inH), int(sr_scale * inW), int(sr_scale * inD)
        )

return out  # N*r*r*r x inC x inH xinW

@staticmethod
def repeat_x(x, scale):
        scale_int = math.ceil(scale)
        N, C, H, W, D = x.size()
        x = x.view(N, C, H, 1, W, 1, D, 1)

        x = torch.cat([x] * scale_int, 3)
        x = torch.cat([x] * scale_int, 5)
        x = torch.cat([x] * scale_int, 7).permute(0, 3, 5, 7, 1, 2, 4, 6)

return x.contiguous().view(-1, C, H, W, D)

@staticmethod
def repeat_weight(weight, scale, inh, inw, ind):
        k = math.ceil(math.pow(weight.size(0), 1 / 3))  #20
outw = inw * scale
        outh = inh * scale
        outd = ind * scale

        weight = weight.view(k, k, k, -1) #20x20x20x
scale_w = (outw + k - 1) // k
        scale_h = (outh + k - 1) // k
        scale_d = (outd + k - 1) // k
        weight = torch.cat([weight] * scale_h, 0)
        weight = torch.cat([weight] * scale_w, 1)
        weight = torch.cat([weight] * scale_d, 2)

        weight = weight[0:outh, 0:outw, 0:outd, :]



return weight’













在 2023-06-14 18:51:06，"XuecaiHu" ***@***.***> 写道：

meta-sr_3d.txt
